software/webapp/src/test/java/brooklyn/test/entity/TestJavaWebAppEntityImpl.java

```diff
@@ -91,4 +73,25 @@ public int getB() {
     public int getC() {
         return c;
     }
+
+    @Override
+    protected SoftwareProcessDriverLifecycleEffectorTasks getLifecycleEffectorTasks() {
+        return new SoftwareProcessDriverLifecycleEffectorTasks() {
+            public void start(java.util.Collection<? extends Location> locations) {
+                ServiceStateLogic.setExpectedState(entity(), Lifecycle.STARTING);
+                LOG.trace("Starting {}", this);
+                entity().setAttribute(SERVICE_PROCESS_IS_RUNNING, true);
+                entity().setAttribute(Attributes.SERVICE_UP, true);
+                ServiceStateLogic.setExpectedState(entity(), Lifecycle.RUNNING);
+            }
+
+            public void stop() {
+                ServiceStateLogic.setExpectedState(entity(), Lifecycle.STOPPING);
+                LOG.trace("Stopping {}", this);
+                entity().setAttribute(Attributes.SERVICE_UP, false);
+                entity().setAttribute(SERVICE_PROCESS_IS_RUNNING, false);
+                ServiceStateLogic.setExpectedState(entity(), Lifecycle.STOPPED);
+            }
+        };
+    }
```

Comment:
Even though it's just a "minor" class, it nonetheless serves as an example of how we recommend navigating this maze of lifecycle methods/effectors/overrides/fields/annotations; doubly so in the absence of decent documentation.